### PR TITLE
OPEN-40 indexing support for testdoubles

### PR DESCRIPTION
**Pull request by anton-larsen-persequor**

### CODE REVIEW

**Comment by MikePSQR** [comment]
https://persequor.atlassian.net/browse/OPEN-83

File: ran-core/src/main/java/io/ran/TestDoubleIndex.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package io.ran;
+
+import javax.inject.Singleton;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class TestDoubleIndex {
+    private Map<String, IndexResult> index = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (MikePSQR) [reply ran-core/src/main/java/io/ran/TestDoubleIndex.java]: also IndexResult

File: ran-core/src/main/java/io/ran/Property.java [hunk 2]
```diff
@@ -151,7 +151,7 @@ public Annotations getAnnotations() {
 	}
 
 	public static class PropertyList extends ArrayList<Property> {
-		private Map<String, Property> propertyMap = Collections.synchronizedMap(new HashMap<>());
+		private ConcurrentMap<String, Property> propertyMap = new ConcurrentHashMap<>();
```
Comment (niels-psqr) [inline ran-core/src/main/java/io/ran/Property.java:154]: Couldn't the left side stay a Map?

File: ran-core/src/main/java/io/ran/TestDoubleDb.java [hunk 3]
```diff
@@ -8,16 +8,22 @@
...
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 @Singleton
 public class TestDoubleDb {
-	public Map<Class<?>, Map<Object, Object>> store = Collections.synchronizedMap(new HashMap<>());
+	public ConcurrentMap<Class<?>, TestDoubleStore<Object, Object>> store = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (MikePSQR) [reply ran-core/src/main/java/io/ran/TestDoubleDb.java]: [54c5092](https://github.com/persequor-com/ran/pull/30/commits/54c5092f1d6512d675fab78fc5a76a15edd046cc)

File: ran-core/src/main/java/io/ran/Property.java (hunk 2 above)
Comment (MikePSQR) [reply ran-core/src/main/java/io/ran/Property.java]: [54c5092](https://github.com/persequor-com/ran/pull/30/commits/54c5092f1d6512d675fab78fc5a76a15edd046cc)

File: ran-core/src/main/java/io/ran/TestDoubleStore.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import javax.inject.Inject;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class TestDoubleStore<K, V> {
+	private final ConcurrentMap<K, V> map = new ConcurrentHashMap<>();
```
Comment (niels-psqr) [inline ran-core/src/main/java/io/ran/TestDoubleStore.java]: Map on the left

Comment (MikePSQR) [reply ran-core/src/main/java/io/ran/TestDoubleStore.java]: [54c5092](https://github.com/persequor-com/ran/pull/30/commits/54c5092f1d6512d675fab78fc5a76a15edd046cc)

File: ran-core/src/test/java/io/ran/IsItsOwnKeyTest.java [hunk 5]
```diff
@@ -22,8 +22,8 @@ public class IsItsOwnKeyTest {
 
 	@Before
 	public void setup() {
-		store = new TestDoubleDb();
 		helper = new GuiceHelper();
+		store = new TestDoubleDb(new MappingHelper(helper.factory));
```
Comment (niels-psqr) [inline ran-core/src/test/java/io/ran/IsItsOwnKeyTest.java:26]: Why not inject it now that its bound in GuideModule anyway (GuiceHelper can instatiate)

File: ran-core/src/test/java/io/ran/TestDoublePerfTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	@Test
+	public void perf() {
+		long s = System.currentTimeMillis();
+		for (int i = 0; i < 1000; i++) {
+			Optional<IndexedCar> foundCar = carRepo.query().eq(IndexedCar::getId, carIds.get(((int) (Math.random() * NUMBER_OF_CARS)))).execute().findFirst();
+			assertTrue(foundCar.isPresent());
+		}
+		System.out.println(System.currentTimeMillis() - s + "ms for pk");
```
Comment (niels-psqr) [inline ran-core/src/test/java/io/ran/TestDoublePerfTest.java:57]: Should we be setting some upper extreme bound for performance? AKA fail if above ??? ms

Comment (MikePSQR) [reply ran-core/src/test/java/io/ran/TestDoublePerfTest.java:59]: we ignore it instead, as we do with performance tests in general

File: ran-core/src/test/java/io/ran/TestDoubleIndexTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+			carRepo.save(car);
+			carIds.add(car.getId());
+		}
+
+	}
+
+	@Test
+	public void useIndexOnPrimaryKeyLookup() {
+		String id = carIds.get(((int) (Math.random() * 4)));
```
[the remark this answers is not in the document]
Comment (MikePSQR) [reply ran-core/src/test/java/io/ran/TestDoubleIndexTest.java:78]: https://www.youtube.com/watch?v=sTMgX1PDGAE

[the remark this answers is not in the document]
Comment (MikePSQR) [reply ran-core/src/test/java/io/ran/TestDoubleIndexTest.java]: [c6cc43e](https://github.com/persequor-com/ran/pull/30/commits/c6cc43e1120f1abf6f6d26589359fdc8c0e93347)

**Review by niels-psqr — APPROVED** [review]
:+1:

**Pull request merged**